### PR TITLE
feat(sns): Data migration that retrofits Swap.{direct_participation_icp_e8s, neurons_fund_participation_icp_e8s} + golden upgrade tests 

### PR DESCRIPTION
**Pull request by aterga**
The PR adds a simple data migration for old Swap canisters that still don't have the fields direct_participation_icp_e8s, neurons_fund_participation_icp_e8s. Additionally, we add golden upgrade tests to check that this migration works as expected in production.

### CODE REVIEW

File: rs/sns/swap/canister/canister.rs [hunk 1]
```diff
@@ -454,6 +454,9 @@ fn canister_post_upgrade() {
     });
 
     init_timers();
+
+    // TODO[NNS1-3386]: Remove this line.
```
Comment (max-dfinity) [inline rs/sns/swap/canister/canister.rs:458]: Probably should not remove it until all existing swaps are migrated to have these fields populated?  Maybe put that in the comment.  

Do we know how many swaps have to be upgraded before we can remove this migration?

File: rs/sns/integration_tests/src/golden_state_swap_upgrade_twice.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        assert_eq!(
+            swap_pre_state, swap_post_state,
+            "Experiment II: Swap state mismatch detected for {}",
+            sns_name
+        );
+    }
+}
+
+#[test]
+fn upgrade_downgrade_swap_boom_dao() {
```
Comment (max-dfinity) [inline rs/sns/integration_tests/src/golden_state_swap_upgrade_twice.rs:138]: "boom_dao" - looks like this test is for all of the SNSes?

**Review by max-dfinity — APPROVED** [review]

File: rs/sns/integration_tests/src/golden_state_swap_upgrade_twice.rs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    {
+        let (mut swap_pre_state, mut swap_post_state) =
+            run_upgrade_for_swap(state_machine, swap_canister_id, swap_wasm_1, sns_name);
+
+        // Some fields need to be redacted as they were introduced after some Swaps were created.
+        redact_unavailable_swap_fields(&mut swap_post_state);
+
+        // Since some SNSs do have (some of) the new fields, we need to redact the same set of
+        // fields from the pre-state, too.
+        redact_unavailable_swap_fields(&mut swap_pre_state);
```
[the remark this answers is not in the document]
Comment (aterga) [reply rs/sns/integration_tests/src/golden_state_swap_upgrade_twice.rs:112]: Yes, this is already covered by Experiment II, in which none of the Swap fields are redacted before they are compared in pre- and post-states.

File: rs/sns/swap/canister/canister.rs (hunk 1 above)
Comment (aterga) [reply rs/sns/swap/canister/canister.rs:458]: > Do we know how many swaps have to be upgraded before we can remove this migration?

I haven't checked, but I could do that tomorrow if we need that info. Note that this migration isn't harmful as it's not doing anything for Swaps that do have the new fields set.

**Pull request merged**